src/nmodl/nocpout.cpp

```diff
@@ -3138,3 +3142,148 @@ Symbol* breakpoint_current(Symbol* s) {
     }
     return s;
 }
+
+// Determine if setdata is required to call FUNCTION or PROCEDURE
+// setdata is required if RANGE var used. For safety, also VERBATIM.
+// Deal with nested calls, via maintaining a list for each func.
+// Note that the nest can be recursive and called function may not
+// yet be defined til entire text is processed.
+
+#include <unordered_set>
+#include <unordered_map>
+
+struct Info {
+    std::unordered_set<Symbol*> func_calls;
+    bool need_setdata{false};
+    bool is_being_looked_at{false};  // avoid recursion loops
+    Item* q{nullptr};                // To be modified if need_setdata.
+};
+
+static std::unordered_map<Symbol*, Info> funcs;
+static Symbol* in_func_;
+
+void check_range_in_func(Symbol* s) {
+    if (in_func_) {
+        // If s is a RANGE variable or nullptr (VERBATIM)
+        // then mark the current function as needing setdata
+        // If s is FUNCTION or PROCEDURE, then add to list
+        Info& i = funcs[in_func_];
+        if (!s) {  // VERBATIM
+            i.need_setdata = true;
+        } else if (s->nrntype & (NRNRANGE | NRNPOINTER)) {
+            i.need_setdata = true;
+        } else if (s->usage & FUNCT) {
+            i.func_calls.insert(s);
+        }
+    }
+}
+
+void set_inside_func(Symbol* s) {
+    in_func_ = s;
+    if (s) {
+        assert(funcs.count(s) == 0);
+        funcs[s] = {};
+    }
+}
+
+// Make sure need_setdata is properly marked for all funcs.
+// I.e on entry, only ones marked are those that use RANGE or VERBATIM.
+// Need to recursively look through func_calls but watch out for loops.
+// If there are no RANGE then VERBATIM is ok and set all need_setdata to false.
+
+static bool check_func(Symbol* s);  // recursive
+
+void func_needs_setdata() {
+    if (!suffix || suffix[0] == '\0') {
```

Comment:
a compiler warning claims that `suffix` is here always evaluated to `true`. Since it's a static char array, that makes sense. Maybe that check is then not needed?

Reply:
You're right. I removed it.